packages/swingset-liveslots/src/cache.js

```diff
@@ -0,0 +1,100 @@
+import { Fail } from '@agoric/assert';
+
+/**
+ * @template V
+ * @callback CacheGet
+ * @param {string} key
+ * @returns {V}
+ */
+
+/**
+ * @template V
+ * @callback CacheSet
+ * @param {string} key
+ * @param {V} value
+ * @returns {void}
+ */
+/**
+ * @callback CacheDelete
+ * @param {string} key
+ * @returns {void}
+ *
+ * @callback CacheFlush
+ * @returns {void}
+ *
+ * @callback CacheInsistClear
+ * @returns {void}
+ */
+/**
+ * @template V
+ * @typedef {object} Cache
+ * @property {CacheGet<V>} get
+ * @property {CacheSet<V>} set
+ * @property {CacheDelete} delete
+ * @property {CacheFlush} flush
+ * @property {CacheInsistClear} insistClear
+ */
+
+/**
+ * Cache of virtual object/collection state
+ *
+ * This cache is empty between deliveries. Within a delivery, the
+ * first access to some data will cause vatstore reads to populate the
+ * cache, then the data is retained until end-of-delivery. Writes to
+ * data will update the cache entry and mark it as dirty. At
+ * end-of-delivery, we flush the cache, writing out any dirty entries,
+ * and deleting all entries.
+ *
+ * This needs RAM for everything read during a delivery (rather than
+ * having a fixed maximum size), but yields a simple (easy to debug)
+ * deterministic relationship between data access and reads/writes to
+ * the backing store.
+ *
+ * @template V
+ * @param {(key: string) => V} readBacking
+ * @param {(key: string, value: V) => void} writeBacking
+ * @param {(key: string) => void} deleteBacking
+ * @returns {Cache<V>}
+ *
+ * This cache is part of the virtual object manager and is not intended to be
+ * used independently; it is exported only for the benefit of test code.
+ */
+export function makeCache(readBacking, writeBacking, deleteBacking) {
+  const stash = new Map();
+  const dirtyKeys = new Set();
+  const cache = {
+    get: key => {
+      if (stash.has(key)) {
+        return stash.get(key);
+      }
+      const value = readBacking(key);
+      stash.set(key, value);
+      return value;
+    },
```

Comment:
I guess we've got insufficient testing for `cache.delete(key); cache.get(key);`, which should return `undefined` without consulting backing storage.
```suggestion
    get: key => {
      if (stash.has(key)) {
        return stash.get(key);
      } else if (dirtyKeys.has(key)) {
        // Respect a pending deletion.
        return undefined;
      }
      const value = readBacking(key);
      stash.set(key, value);
      return value;
    },
```

Reply:
Oops, excellent catch, yeah the VOM data cache doesn't ever `get` a key that's not supposed to exist (it only `delete`s after all means to get a Representative are gone), and the context cache never deletes or uses `writeBacking`. I'll update the implementation and the test.